istioctl/pkg/waypoint/waypoint.go

```diff
@@ -49,33 +50,63 @@ var (
 
 	deleteAll bool
 
-	trafficType       string
+	trafficType       = constants.ServiceTraffic
 	validTrafficTypes = sets.New(constants.ServiceTraffic, constants.WorkloadTraffic, constants.AllTraffic, constants.NoTraffic)
 
-	waypointName string
+	waypointName      = constants.DefaultNamespaceWaypoint
+	enrolledNamespace = constants.DefaultEnrolledNamespace
 )
 
 const waitTimeout = 90 * time.Second
 
 func Cmd(ctx cli.Context) *cobra.Command {
-	makeGatewayName := func(name string) string {
-		if name == "" {
-			name = constants.DefaultNamespaceWaypoint
+	annotateNamespaceWithWaypoint := func(kubeClient kube.CLIClient) error {
+		nsObj, err := kubeClient.Kube().CoreV1().Namespaces().Get(context.Background(), enrolledNamespace, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("namespace: %s not found", enrolledNamespace)
+		} else if err != nil {
+			return fmt.Errorf("failed to get namespace %s: %v", enrolledNamespace, err)
 		}
-		return name
+		if nsObj.Annotations == nil {
+			nsObj.Annotations = map[string]string{}
+		} else if _, exists := nsObj.Annotations[constants.AmbientUseWaypoint]; exists && enrolledNamespace == constants.DefaultEnrolledNamespace {
```

Comment:
Not sure about this. What if this is set and the user wants to set it back to "default" namespace? This won't allow them to do that. Should we allow that? @howardjohn

Reply:
Maybe this isn't relevant since makeGateway is only called during generating or applying a wapoint configuration?

Reply:
But I can imagine a possible case where a namespace already has a waypoint enrolled. Should we overwrite the existing one, allow for multiple waypoints be enrolled for one namespace or just let the user know that a waypoint is already enrolled for this namespace and don't overwrite it? cc/ @ilrudie

Reply:
I'm a -1 on having this be a default value. Traffic redirection through a waypoint shouldn't be done lightly or accidentally; users can provide their namespace in the flag IMO

Reply:
I do think it would make the code a little more simpler to not have the default. Default value makes it hard to discern whether or not it was not set or set to default by the user or set to default by the logic. 

As an alternative to the default, keeping in mind we want some decent UX around the case where it's not set, wdyt about adding a warn log for when it's noticed that it's empty (a.k.a. not set) and just not annotating at all in that case. Then only annotate the namespace when a user tells us to. cc/ @keithmattix ?

Reply:
I don't think I understand this. AFAIK this is checking if its `default`, referring to a _Namespace_.

But a plain string is referring to a waypoint name, not the namespace?

Reply:
Like below we use DefaultNamespaceWaypoint

Reply:
I'm not sure I understand your question.

But right now the logic is,

- default values for `waypointName` and `enrolledNamespace`, which reference those constants (`DefaultNamespaceWaypoint` and `DefaultEnrolledNamespace`) are:
```
waypointName = "default-namespace"
enrolledNamespace = "default"
```
(*Maybe waypointName should have a different default other than `default-namespace`. That does seem a bit confusing. Maybe DefaultWaypointName with a different string value without namespace in it.*)

- if a user uses the `--name` flag and passes a string that is not `""`, it will overwrite the `waypointName` default value of `"default-namespace"` otherwise that is the name the waypoint will be given when created.

- if a user uses the `--enroll-namespace` flag and passes a string that is not `""`, it will overwrite the `enrolledNamespace` value of `default`.
    - then, when makeGateway is called it will call `annotateNamespaceWithWaypoint` and add that annotation on the namespace object. Telling it to use the waypoint (`istio.io/use-waypoint=<waypointName-value>`). If there is already a use waypoint annotation set and the value of  `enrolledNamespace` is still the default value, then don't annotate the namespace by returning nil. Otherwise, set the annotation to waypointName which will be either "default-namespace" or the value the user could have set. This should happen when enrolledNamespace is provided or annotation is not set.

Reply:
IMO enroll-namespace is just on or off. If it's on it should (try to) enroll the namespace that the waypoint is deployed in.

Reply:
So I would think we have 

```
istioctl x waypoint apply
istioctl x waypoint apply --name=foo
istioctl x waypoint apply --name=foo --enroll-namespace # annotate with istio.io/use-waypoint=foo
```

?

Reply:
Ok, so a `bool` and not a `string`. And if on then we just use the namespace for this waypoint and if off, do nothing?

Reply:
yeah, that would be my preference I think. keep it pretty simple